### ISSUE TITLE
Asking if this is a problem on newer QF: When doing file-into or go-to of a non-terminal folder path, QF will not allow repeat actions

### ISSUE THREAD

**Issue by jaysmithgit** [RealRaven2000/QuickFolders#240]
I do *NOT* expect a solution to this problem on my "old" QF 5.5.2 PRO on TB 78.7.1 (64-bit) on Ubuntu Linux 20.04.  However, I am hoping to find out if this is still a problem on later versions of QF/TB so that I will be prepared when I upgrade.

My TB/QF is running directly, locally on a very fast workstation with a lot of memory. I am connecting to a IMAP server (local on our premises), over wired gigabit ethernet from that workstation. The server running the IMAP does have a lot of other responsibilities, but it does have 468 GB of RAM.  We currently have about 9950 mail folders on the IMAP server, in a hierarchical folder structure that is up to about 8 levels deep.

The problem I am describing happens *every* the time in this situation, when I try to repeat the same action.

Here is are some example folder names (fake), several levels deep in the structure.  All our actual mail-message-containing folders are terminal folders; there is no mixing of folders and messages in the the same folder.] 

`.../.../vendors/banks/bank_big/savings_account/correspondence`
`.../.../vendors/banks/bank_big/savings_account/transactions`
`.../.../vendors/banks/bank_small/checking_account/correspondence`
`.../.../vendors/banks/bank_small/checking_account/transactions`

Because there are many mail-containing folders with the (terminal) name "correspondence", when I am typing text for file-into-folder or go-to-folder, will instead type 

If the first character I type (to file-into or to go-to) is "b", it is extremely fast.
If the first character I type (to file-into or to go-to) is "e", it is extremely fast.

... if the first character I type (to file-into or to go-to) is *ANY* letter or number, *EXCEPT* the number "1", it is still extremely fast.

BUT, if the first character I type (to file-into or to go-to) is the NUMBER "1", it is very slow. If the first number I type is a "1", then the process can stall for 10-30 seconds.  Even when I try to type very fast the next digits, it still stalls. 

60% of our folder names include the number "1" somewhere in the folder name, thus that is about 6000 folders.

I am guessing that...

a) The QF function is *immediately* trying to read the index, or do indexing-on-the-fly, all the occurrences of the first character typed. I suspect that if there was a few milliseconds of delay to listen for additional characters, it could help. I do not know how the indexing works or how QF reads the index (character at a time, or strings), but if there was time allowed for the user to enter two or three characters before QF starts working on it, then it might solve the problem.

b) There has not been a large change in the quantity of our folders between the time that the number "1" was fast and when the number "1" slowed down.  However, naturally over time TB's "database" files grow. TB 78 is tremendously more stable than the ancient version I was using before moving to 78. In the old days, I had to remove the "database" files every 6-12 months so that TB could function again.  However, that was extremely inconvenient -- it took almost a day to rebuild the folder indexes (for each user's TB) so that the folders were "visible" to our old filing system. (That was before we switched to QF, but QF still requires a similar process of making the folders "visible" all the way down to hierarchical levels).  HOWEVER, I am wondering if the "database" files are now too big for QF, or if there needs to be an increase some sort of memory/process allotment, etc.  Here is a list of the (largest-first) largest "database" files for my TB profile:

`7257814 Mar  3 19:13 training.dat`
`6911134 Mar  5 12:59 panacea.dat`
`5509088 Mar  5 12:50 abook.sqlite-wal`
`5242880 Oct  9 16:15 favicons.sqlite`
`5242880 Mar  4 16:46 places.sqlite`
`2426640 Mar  4 19:28 places.sqlite-wal`
`1245184 Jan 28 13:16 global-messages-db.sqlite`
` 524288 Apr  3  2021 blist.sqlite`

Does anybody have a similar problem either with QF 5.5.2 or newer versions?  (Does anybody have a similar IMAP environment, quantity of folders, etc.?)

Jay Smith

**Comment by RealRaven2000** [comment]
seems very similar to #239 ?

**Comment by jaysmithgit** [comment]
I do *NOT* expect a solution to this problem on my "old" QF 5.5.2 PRO on TB 78.7.1 (64-bit) on Ubuntu Linux 20.04.  However, I am hoping to find out if this is still a problem on later versions of QF/TB so that I will be prepared when I upgrade.

My TB/QF is running directly, locally on a very fast workstation with a lot of memory. I am connecting to a IMAP server (local on our premises), over wired gigabit ethernet from that workstation. The server running the IMAP does have a lot of other responsibilities, but it does have 48 GB of RAM.  We currently have about 9950 mail folders on the IMAP server, in a hierarchical folder structure that is up to about 8 levels deep.

The problem I am describing happens *every* the time in this situation, when I try to repeat the same action using the keyboard.

Here is are some example folder names (fake), several levels deep in the structure.  All our actual mail-message-containing folders are terminal folders; there is no mixing of folders and messages in the the same folder.] 

`.../.../vendors/banks/bank_big/savings_account/correspondence`
`.../.../vendors/banks/bank_big/savings_account/transactions`
`.../.../vendors/banks/bank_small/checking_account/correspondence`
`.../.../vendors/banks/bank_small/checking_account/transactions`

Because there are many mail-containing folders with the (terminal) folder name "correspondence", when I am typing text for the file-into-folder or go-to-folder functions, I will instead type "bank_big" and then navigate in the resulting list to select `.../.../vendors/banks/bank_big/savings_account/correspondence`

However, if I do it that way...

1) QF *is* aware that the last action I did was for
`.../.../vendors/banks/bank_big/savings_account/correspondence`

2) However, QF will *NOT* allow me to repeat that action, using the KEYBOARD, or to do the other action (go-to or file-into), UNLESS I start over and do the whole process again (typing "bank_big").  Even though `.../.../vendors/banks/bank_big/savings_account/correspondence` is in the list displayed, I can't use the KEYBOARD actions to file-into it or go-to it unless I start all over.

But... If instead, the first time I had typed part or all the **TERMINAL** FOLDER NAME `correspondence` and selected the desired folder from a (long) list of possibilities, then QF *DOES* allow me to use the KEYBOARD to repeat the action or the other (go-to or file-into) action.

The difference is:

A) First typing part of a **NON-TERMINAL** folder path: Function works the first time, but cannot be repeated from the KEYBOARD without starting over).  (This means typing something *OTHER than* all or part of "correspondence".)

B) First typing all or part of a **TERMINAL** folder name (i.e. all or part of "correspondence"): Function works the first and allows me to repeat the action (doing file-into and/or go-to) as many times as I want using the KEYBOARD.


Does anybody have a similar problem either with QF 5.5.2 or newer versions?  (Does anybody have a similar IMAP environment, quantity of folders, etc.?)

Again, I am NOT trying to get this fixed for QF 5.5.2. However, I would like to find out if the same problem exists in the more recent versions.

Jay Smith

**Comment by jaysmithgit** [comment]
> seems very similar to #239 ?

While I do not know enough about the internal working of these processes, I do **not** think that this is related to #239.

**Comment by jaysmithgit** [comment]
I failed to mention that QF **will** allow me to drag and drop to the item in Recent Folders in these (and all) situations.  However, it just won't allow me to do repeat the function from the keyboard.

**Comment by RealRaven2000** [comment]
If it might help you - after invoking quickMove (either by dragging mails onto the button or typing Shift+M) you can enter the character "=" as first thing and it will pop up the recent folders menu. Of course this is only useful if you know that you did a quickJump / quickMove to the wanted folder in the last 10 moves. I am yet waiting for a intuitive and fast way of integrating the typed results list and recent folders into the same menu. Ideally both the results and the recent list would display at the same time, but on top of each other would likely be a very long and not useful menu (especially after only typing a single character).

**Comment by jaysmithgit** [comment]
Assuming that the character you are suggesting to use is the "=" [equals] sign, and assuming you meant that typing "=" would open the recent folders list.... that does not function for me in 5.5.2. PRO.  

Perhaps I am misunderstanding something. I have tried just "=" and "=" with various combinations, but nothing seems to happen.

**Comment by RealRaven2000** [comment]
> Assuming that the character you are suggesting to use is the "=" [equals] sign, and assuming you meant that typing "=" would open the recent folders list.... that does not function for me in 5.5.2. PRO.
> 
> Perhaps I am misunderstanding something. I have tried just "=" and "=" with various combinations, but nothing seems to happen.

You should probably try a current version (5.8.3) i you're already using Thunderbird 78 or later. remember you can always roll back by downloading older versions from here:
https://addons.thunderbird.net/en-US/thunderbird/addon/quickfolders-tabbed-folders/versions/

**Comment by RealRaven2000** [comment]
> Because there are many mail-containing folders with the (terminal) folder name "correspondence", when I am typing text for the file-into-folder or go-to-folder functions, I will instead type "bank_big" and then navigate in the resulting list to select .../.../vendors/banks/bank_big/savings_account/correspondence

By the way, did you try typing:

bank_b>corr  

the "greater than" character `>` can be used to skip parent folders (and the order of parent folders does not matter, so typing 

sav>ban>corr   

should also find the folder. Clicking on the (?) button beside the search field gives a brief overview on the syntax of the search field.

![image](https://user-images.githubusercontent.com/6314674/157093432-f0383465-7644-49a4-a958-4b040f9c294a.png)

**Comment by jaysmithgit** [comment]
Okay, I installed 5.8.3. I had to restart TB for the "=" function to work.  

So, now if I do "s" [for ME "save/move"] [or "g" for ME "go to"] and then "=" it DOES open up the recent list and I can arrow-key-navigate up/down the list.

BUT... it is **not** working as expected.  Instead the ["s", "=", and navigating in the recent list, "Enter" to select an item in the recent list}  is NOT moving the email message.  Instead it is going [jumping] to and opening the target folder.

[Doing the testing also revealed another problem in 5.5.2 that I had not been aware are, but I have to test more in 5.8.3 to determine if the problem still exists. It is complicated.]

Do you see the same behavior or it is just me?

**Comment by RealRaven2000** [comment]
> BUT... it is **not** working as expected. Instead the ["s", "=", and navigating in the recent list, "Enter" to select an item in the recent list} is NOT moving the email message. Instead it is going [jumping] to and opening the target folder.
> 
> Do you see the same behavior or it is just me?

Yes - I think this may be a bug. It should definitely move the mail. The `=` shortcut was never officially documented because of the developer of nostalgy++ complaining that too many features of his Add-on were demanded by users from QuickFolders. So on top of integrating Recent folders with quickMove / quickJump  in ajn intuitive, / easy way I also have to worry about taking away users from his base which is not something I want.

**Comment by jaysmithgit** [comment]
I hope that you will choose to fix this (possible) bug.

I have found another bug which I will report on a different issue.

**Comment by RealRaven2000** [comment]
> I hope that you will choose to fix this (possible) bug.

I will, if you like you can add another issue for this if I forget to do it over the next couple of days.

**Comment by RealRaven2000** [comment]
I will fix the issue with the `=` character, but let me just circle back to the original problem:

when you invoke quickJump (with a keyboard shortcut, it doesn't matter which one you have configured) or quickMove (by dragging an email to the button or through the configured shortcut), QuickFolders tries to be helpful by suggesting the last successful folder. The idea is to accept this suggestion using the Enter key or to start typing a different folder name - hence the text is highlighted to allow simply overwriting the suggestion:

![image](https://user-images.githubusercontent.com/6314674/157210139-e45455c2-4996-4660-8c4e-a50d9e79d750.png)

I do not want to show the suggestions menu at this stage to keep the interface as uncluttered as possible. This also means that you cannot use the arrow down key in order to bring up suggestions, you have to edit the existing string to do this. For instance you could delete the last character of the search text, or enter a space afterwards. Unfortunately just using the cursor keys in the text will not bring up the suggestions. So maybe there is another way to do this in an intuitive manner. **Maybe cursor down should bring up the menu with suggested folders.**

**Comment by RealRaven2000** [comment]
Testing the new version (5.9 pre 28)  I just uploaded at #241, I did this test:

Trying to reach the folder Local1/Dev/quickFilters, I used the quickJump shortcut (default, Shift+J, the actual shortcut doesn't matter though) and typed this:

![image](https://user-images.githubusercontent.com/6314674/157696763-9b3dfbf1-2486-4819-beff-49a58c145b72.png)

I then selected the top item (which now forces QF to remember the full folder path URI in the preference setting
`extensions.quickfolders.quickMove.lastFolderURI` which in this case has the value "mailbox://nobody@Local%20Folders/Local1/Dev/quickFilters"

when I now enter Shift+J the search box displays the last entry _without_ a list of suggestions:
![image](https://user-images.githubusercontent.com/6314674/157697431-fd573b6f-8c81-47ad-a4ad-65ff0f6acdb3.png)

Immediately hitting Enter will jump into the last folder

**Comment by RealRaven2000** [comment]
Just for clarification, this item is about repeating the last Jump using the "Enter" key, is this correct?

**Comment by jaysmithgit** [comment]
> Just for clarification, this item is about repeating the last Jump using the "Enter" key, is this correct?

I think that if I attempt to answer this, it will just make for more confusion because I am now using the new (NON-beta) QF and thus various things have changed.

The work you have done and are about to do ( #242 ) have and will change things further.

I have yet to test the latest beta you provided (I cannot do that until at least tonight or tomorrow).

I think everything is on the right track, and as soon as #242 is done, I think everything will be done and wonderful.

If there is/are any issue(s) remaining after #242 is done, then I think I should start a new thread which will refer only to the latest (beta) version.

Jay

**Comment by RealRaven2000** [comment]
> BUT... it is **not** working as expected. Instead the ["s", "=", and navigating in the recent list, "Enter" to select an item in the recent list} is NOT moving the email message. Instead it is going [jumping] to and opening the target folder.

I posted a version for testing on a new issue that's dedicated to the recent folders shortcut:

https://github.com/RealRaven2000/QuickFolders/issues/242#issuecomment-1065524744

**Comment by fanis** [comment]
I came here searching as I also have same folder names in differently named subfolders and subsequent go/save/copy just selects the first in the list.

I tried QuickFolders-mx-5.9pre33.zip and it works wonderfully, nice to see such commitment and professionalism yet again @RealRaven2000 , and great job reporting and working on it @jaysmithgit .
